File: #72_Edit_Distance/Solution.java
public class Solution {
    public int minDistance(String word1, String word2) {
        if (word1 == null || word1.length() == 0) return (word2 == null) ? 0 : word2.length();
        if (word2 == null || word2.length() == 0) return word1.length();
        // dp[i][j] minimum distance between w1[0, i) and w2[0, j)
        int[] dp = new int[word2.length() + 1];
        for (int j = 1; j <= word2.length(); j++) {
        	dp[j] = j;
        }
        for (int i = 1; i <= word1.length(); i++) {
			int prev = dp[0];
			dp[0] = dp[0] + 1;
        	for (int j = 1; j <= word2.length(); j++) {
    			int tmp = dp[j];
        		if (word1.charAt(i - 1) == word2.charAt(j - 1)) dp[j] = prev;
        		else dp[j] = Math.min(Math.min(prev + 1, dp[j] + 1), dp[j - 1] + 1);
        		prev = tmp;
        	}
        }
        return dp[word2.length()];
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	System.out.println(sol.minDistance("sea", "ate"));
    }
}

File: #72_Edit_Distance/Solution.py
'''
DP Solution
dp[i][j]: min step for word1[0, i) and word2[0, j) to transform
word1[i] = c
word2[j] = d
if c == d:
	dp[i][j] = dp[i - 1][j - 1]
else:
	convert c to d: dp[i][j] = dp[i - 1][j - 1] + 1
	insert d after c: dp[i][j] = dp[i][j - 1] + 1
	delete c: dp[i][j] = dp[i - 1][j] + 1

time: O(m * n)
space: O(n)
'''

class Solution:
    # @param {string} word1
    # @param {string} word2
    # @return {integer}
    def minDistance(self, word1, word2):
        dp = [[j + i for j in range(len(word2) + 1)] for i in range(len(word1) + 1)]
        for i in range(1, len(word1) + 1):
        	for j in range(1, len(word2) + 1):
        		if word1[i - 1] == word2[j - 1]:
        			dp[i][j] = dp[i - 1][j - 1]
        		else:
        			dp[i][j] = min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]) + 1
        return dp[len(word1)][len(word2)]

sol = Solution()
print sol.minDistance("intention", "execution")